log4j-to-jul/src/main/java/org/apache/logging/log4j/tojul/LazyLog4jLogRecord.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache license, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.tojul;
+
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import org.apache.logging.log4j.util.StackLocatorUtil;
+
+/**
+ * Extension of {@link java.util.logging.LogRecord} with lazy get source related methods based on Log4j's {@link StackLocatorUtil#calcLocation(String)}.
+ */
+final class LazyLog4jLogRecord extends LogRecord {
+
+    private static final long serialVersionUID = 6798134264543826471L;
+
+    // parent class LogRecord already has a needToInferCaller but it's private
+    private transient boolean inferCaller = true;
+
+    private final String fqcn;
+
+    LazyLog4jLogRecord(String fqcn, Level level, String msg) {
+        super(level, msg);
+        this.fqcn = fqcn;
+    }
+
+    @Override
+    public String getSourceClassName() {
+        if (inferCaller) {
+            inferCaller();
+        }
+        return super.getSourceClassName();
+    }
+
+    @Override
+    public String getSourceMethodName() {
+        if (inferCaller) {
+            inferCaller();
+        }
+        return super.getSourceMethodName();
+    }
+
+    private synchronized void inferCaller() {
```

Comment:
I don't think synchronized is necessary based on how this class is used and the thread-safety of the parent class. Synchronization here doesn't buy us anything because it only guards the method itself, not the `inferCaller` check elsewhere (which isn't volatile, nor do I think it should be) so it would only prevent additional concurrent calls from running in parallel.

Reply:
@garydgregory @carterkozak can I let you guys hash this out among ourselves and let me know your conclusion? I'm happy if you are (both) happy, but I'd prefer not going back and forth a few times between you... 😸  I believe this comment appears to request to (re-)remove `synchronized` (again, after @garydgregory suggested it be added), and also for `volatile` to be removed from `inferCaller`. Please both ACK?

Reply:
s/ourselves/yourselves/

Reply:
Let's remove the `synchronized` keyword please, sorry for the churn!

>  and also for volatile to be removed from inferCaller

I don't think `inferCaller` is volatile, no change needed there!

Reply:
OK, `synchronized` removed again now.

> I don't think `inferCaller` is volatile, no change needed there!

Duh, on TGIF the `transient` morphed into a `volatile` in my mind 😈